### ISSUE TITLE
pnpm publish cannot_resolve_workspace_protocol

### ISSUE THREAD

**Issue by VividLemon** [pnpm/pnpm#5094]
There are other similar issues it seems, but they are somewhat different
### pnpm version:
7.6.0
### Code to reproduce the issue:
Create a workspace,
have one workspace have a dependency of another, in my case it was one that is for a tsconfig, to which my main package has a dependency of.
Run pnpm publish,
### Expected behavior:
It should work fine
### Actual behavior:
`Cannot resolve workspace protocol of dependency "tsconfig" because this dependency is not installed. Try running "pnpm install".` Even though pnpm says already up to date when doing pnpm i, and it is viewable in the node_modules folder for that workspace package.
### Additional information:

 - `node -v` prints:v16.14.2
 - Windows, macOS, or Linux?: Windows 11

At this point it prevents me from publishing, so it's a bit of a tough issue.

**Comment by VividLemon** [comment]
Okay, so I tried something that I think may have caused an issue. Not including a version number in the dependent (tsconfig), causes this failure. By adding a version in the package.json, the error stopped and could run. When a collaborator sees this, please close the issue. I'm just keeping it open so it may be known that not having the version was the issue. Perhaps to help a future issue haver, or to possibly talk about potential fixes like a better error message, or note in the error, or a note in docs.

**Comment by trappar** [comment]
I'm getting this same issue. I'm using `changesets` and I've removed the version from some packages as a way to prevent those ones from being versionable. As soon as a single package is missing a version, PNPM fails with the error when trying to publish:
```
ERR_PNPM_CANNOT_RESOLVE_WORKSPACE_PROTOCOL  Cannot resolve workspace protocol of dependency "@curology/eslint-config-typescript" because this dependency is not installed. Try running "pnpm install".
```

It does this even when running publish on a single specific package which _**does**_ have a version. I guess I might just use NPM to publish for now...?

**Comment by vaynevayne** [comment]
+1

**Comment by MikeYermolayev** [comment]
+1. Trying to set up `changesets` I removed "version" field from packages that are not intended for publish.

@trappar did you find any workaround maybe? Thanks a lot for any tip

**Comment by dipunm** [comment]
I have come across this issue while attempting to use [changesets](https://github.com/changesets/changesets) with `pnpm`.

The issue occurs when using private packages for configuration purposes (`eslint`, `tsconfig` etc). By removing the `"version"` field, changesets will understand that this package is not a versionable package and will not offer to bump its version. This is typically desired for private internal packages.

`pnpm` does not know what to do with these workspace dependencies if there is no version attached to their package.json. Either changesets needs to include an alternative way to ignore private workspace-only packages, or pnpm needs to understand that a package without a version should probably remove the entry at the time of publish.

**Comment by alexander-dermicus** [comment]
Changing "workspace:*" to "workspace:x" fixed the issue for me.

**Comment by alexander-dermicus** [comment]
Adding `auto-install-peers=true` in the `.npmrc` file also fixed the issue.

**Comment by gxxcastillo** [comment]
> I guess I might just use npm publish instead for now 😞

I'm also using changesets and seeing this error, all my private, internal, packages are hardcoded to version: 0.0.0. Using `npm publish` was the only way I could publish.

**Comment by Eldow** [comment]
One solution I found was to explicitly convert the `workspace:*` to their respective versions before calling `pnpm publish`:

```ts
// tools/scripts/create-exportable-manifest.mjs
import { createExportableManifest } from '@pnpm/exportable-manifest';
import { readProjectManifestOnly } from '@pnpm/read-project-manifest';
import { writeFile } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);

const __dirname = path.dirname(__filename);

const [, , name] = process.argv;

const projectDir = path.join(__dirname, `../../packages/${name}`); // folder of "package.json" to be translated
const distDir = path.join(__dirname, `../../dist/${name}`); // folder to save the translated one

(async () => {
  const manifest = await readProjectManifestOnly(projectDir);
  const exportable = await createExportableManifest(projectDir, manifest);
  await writeFile(
    `${distDir}/package.json`,
    JSON.stringify(exportable, undefined, 2)
  );
})()
  .then(() => {})
  .catch(console.error);
  ```
  Feel free to configure the `projectDir` and `distDir` variables according to your repo structure.
  You can call it like this `node tools/scripts/create-exportable-manifest.mjs packageName`, it will replace the package.json in your dist folders with one containing the actual versions of your packages inside the workspace. Without the `workspace:*` versions, your `pnpm publish` should work.

**Comment by millsp** [comment]
For those who found this is caused by a private package, this helped me
https://github.com/pnpm/pnpm/issues/4164#issuecomment-1236762286

tldr; You need to set a version even on a private package, even if it's a local dev dependency

**Comment by altano** [comment]
So to summarize the conflict that led some of us here:

pnpm: requires version, even on private packages [^1]
changeset: requires no version, even on private packages, or it will recommend version bumping them [^2]

I believe the best workaround is to use the undocumented config option in changesets to tell it to ignore all private packages, regardless of version: just add `"privatePackages": false` to config.json. Then pnpm will be able to publish and changesets will ignore your private packages. This worked for me.

[^1]: I assume this is a bug or limitation in pnpm, and not by design
[^2]: The changesets author is considering changing the default behavior to ignore all private packages: https://github.com/changesets/changesets/blob/0bf89b3709e3e3df6ed5dbb8ece0fb000a55d5f4/packages/config/src/index.ts#L482

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1: `pnpm publish --dry-run` of a workspace project whose dependencies use the `workspace:` protocol succeeds; the protocol is resolved in the published manifest.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
